test/api/route/messages_test.dart

```diff
@@ -436,4 +451,196 @@ void main() {
+          anchor: AnchorCode.oldest,
+          numBefore: 0, numAfter: 20,
+          narrow: const AllMessagesNarrow().apiEncode(),
+          op: UpdateMessageFlagsOp.add, flag: MessageFlag.read,
+          expected: {
+            'anchor': 'oldest',
+            'num_before': '0',
+            'num_after': '20',
+            'narrow': jsonEncode([]),
+            'op': 'add',
+            'flag': 'read',
+          });
+      });
+    });
+
+    test('narrow uses resolveLegacyElements to encode', () {
+      return FakeApiConnection.with_(zulipFeatureLevel: 176, (connection) async {
+        connection.prepare(json: mkResult(foundOldest: true).toJson());
+        await checkUpdateMessageFlagsForNarrow(connection,
+          anchor: AnchorCode.oldest,
+          numBefore: 0, numAfter: 20,
+          narrow: [ApiNarrowDm([123, 234])],
+          op: UpdateMessageFlagsOp.add, flag: MessageFlag.read,
+          expected: {
+            'anchor': 'oldest',
+            'num_before': '0',
+            'num_after': '20',
+            'narrow': jsonEncode([
+              {'operator': 'pm-with', 'operand': [123, 234]},
+            ]),
+            'op': 'add',
+            'flag': 'read',
+          });
+      });
+    });
+
+    test('numeric anchor', () {
+      return FakeApiConnection.with_((connection) async {
+        connection.prepare(json: mkResult(foundOldest: false).toJson());
+        await checkUpdateMessageFlagsForNarrow(connection,
+            anchor: const NumericAnchor(42),
+            numBefore: 0, numAfter: 20,
+            narrow: const AllMessagesNarrow().apiEncode(),
+            op: UpdateMessageFlagsOp.add, flag: MessageFlag.read,
+            expected: {
+              'anchor': '42',
+              'num_before': '0',
+              'num_after': '20',
+              'narrow': jsonEncode([]),
+              'op': 'add',
+              'flag': 'read',
+            });
+      });
+    });
+  });
+
+  group('markAllAsRead', () {
+    Future<void> checkMarkAllAsRead(
+      FakeApiConnection connection, {
+      required Map<String, String> expected,
+    }) async {
+      connection.prepare(json: {});
+      await markAllAsRead(connection);
+      check(connection.lastRequest).isNotNull().isA<http.Request>()
+        ..method.equals('POST')
+        ..url.path.equals('/api/v1/mark_all_as_read')
+        ..bodyFields.deepEquals(expected);
+    }
+
+    test('smoke', () {
+      return FakeApiConnection.with_((connection) async {
+        await checkMarkAllAsRead(connection, expected: {});
+      });
+    });
+  });
+
+  group('markStreamAsRead', () {
+    Future<void> checkMarkStreamAsRead(
+      FakeApiConnection connection, {
+      required int streamId,
+      required Map<String, String> expected,
+    }) async {
+      connection.prepare(json: {});
+      await markStreamAsRead(connection, streamId: streamId);
+      check(connection.lastRequest).isNotNull().isA<http.Request>()
```

Comment:
```suggestion
      check(connection.lastRequest).isA<http.Request>()
```

As I mentioned in our call yesterday, the `isNotNull` is redundant with the subsequent `isA`. The type we're checking for with `isA` already excludes null.

(I think I originally introduced this redundant pattern to this file. But we should stop propagating it.)

Reply:
→ #369 for the existing cases of that pattern.